auslib/test/test_db.py

```diff
@@ -637,6 +638,376 @@ def testHistoryGetChangeWithDataVersionWithNonPrimaryKeyColumn(self):
+        what = {"fooid": 3, "bar": "blah", "when": 343, "data_version": 1}
+        self.assertRaises(PermissionDeniedError, self.sc_table.insert, changed_by="nancy", **what)
+
+    def testUpdateNoChangesSinceCreation(self):
+        where = [self.sc_table.sc_id == 1]
+        what = {"when": 888, "foo": "bb"}
+        self.sc_table.update(where, what, changed_by="bob", old_data_version=1)
+        row = self.sc_table.t.select().where(self.sc_table.sc_id == 1).execute().fetchall()[0]
+        history_row = self.sc_table.history.t.select().where(self.sc_table.history.sc_id == 1).execute().fetchall()[0]
+        self.assertEquals(row.scheduled_by, "bob")
+        self.assertEquals(row.when, 888)
+        self.assertEquals(row.data_version, 2)
+        self.assertEquals(row.base_fooid, 1)
+        self.assertEquals(row.base_foo, "bb")
+        self.assertEquals(row.base_bar, "barbar")
+        self.assertEquals(row.base_data_version, 1)
+        self.assertEquals(history_row.changed_by, "bob")
+        self.assertEquals(history_row.scheduled_by, "bob")
+        self.assertEquals(history_row.when, 888)
+        self.assertEquals(history_row.data_version, 2)
+        self.assertEquals(history_row.base_fooid, 1)
+        self.assertEquals(history_row.base_foo, "bb")
+        self.assertEquals(history_row.base_bar, "barbar")
+        self.assertEquals(history_row.base_data_version, 1)
+
+    def testUpdateNoChangesSinceCreationWithDict(self):
+        where = {"sc_id": 1}
+        what = {"when": 888, "foo": "bb", "data_version": 1, "fooid": 1}
+        self.sc_table.update(where, what, changed_by="bob", old_data_version=1)
+        row = self.sc_table.t.select().where(self.sc_table.sc_id == 1).execute().fetchall()[0]
+        history_row = self.sc_table.history.t.select().where(self.sc_table.history.sc_id == 1).execute().fetchall()[0]
+        self.assertEquals(row.scheduled_by, "bob")
+        self.assertEquals(row.when, 888)
+        self.assertEquals(row.data_version, 2)
+        self.assertEquals(row.base_fooid, 1)
+        self.assertEquals(row.base_foo, "bb")
+        self.assertEquals(row.base_bar, "barbar")
+        self.assertEquals(row.base_data_version, 1)
+        self.assertEquals(history_row.changed_by, "bob")
+        self.assertEquals(history_row.scheduled_by, "bob")
+        self.assertEquals(history_row.when, 888)
+        self.assertEquals(history_row.data_version, 2)
+        self.assertEquals(history_row.base_fooid, 1)
+        self.assertEquals(history_row.base_foo, "bb")
+        self.assertEquals(history_row.base_bar, "barbar")
+        self.assertEquals(history_row.base_data_version, 1)
+
+    def testUpdateWithBadConditions(self):
+        where = [self.sc_table.sc_id == 1]
+        what = {"telemetry_product": "boop", "telemetry_channel": "boop", "telemetry_uptake": 99}
+        self.assertRaisesRegexp(ValueError, "Invalid combination of conditions", self.sc_table.update, where, what, changed_by="bob", old_data_version=1)
```

Comment:
Yeah. In order to make sure it's not possible for the conditions to get into an invalid state, you need to remove "when" and add the telemetry ones. Is that awkward to you? An alternative here could be not to allow the type of condition to change (require the user to remove the old scheduled change and add a new one).